### Invers Multiplikatif

In [2]:
from math import nan

# the function returns k^{-1} (mod n), if exists
def find_mult_inverse(k, n) -> list[int]:
    a = [nan, k]
    b = [nan, n]
    q = [nan, k//n]
    r = [nan, k%n]
    s = [1, 0]
    t = [0, 1]

    i = 1
    while (r[i]!=0):
        i+=1
        a.append(b[i-1])
        b.append(r[i-1])
        q.append(a[i] // b[i])
        r.append(a[i] % b[i])
        
        s.append(s[i-2]-(s[i-1]*q[i-1]))
        t.append(t[i-2]-(t[i-1]*q[i-1]))
        
    # this is the case when the GCD is not 1
    if b[i] != 1:
        print("Multiplicative inverse does not exist!")
        return -1

    return s[i] % n


P = 1_000_000_007
test = [(i, P) for i in range(345678, 987654+1)]

result = [find_mult_inverse(k, n) for k, n in test]
result = sum(result)
result = result % P

print("Result: ")
print(result)


Result: 
148059137


### Logaritma Diskrit

In [1]:
from time import perf_counter


def discrete_log(a, n, b) -> int:
    # this function returns log_{a, n} (b), i.e. 
    # the smallest integer k satisfying a^k ≡ b (mod n)
    # if such k does not exist, print "logarithm does not exist"
    k = 1
    while True:
        mod = a**k % n
        if mod == b:
            break
        
        if k > n:
            return 0
        
        k += 1
    
    return k


b = 103
unexist = 0
sum_result = 0

start = perf_counter()
for n in range(4567, 8765 + 1):
    a_sum = 0
    
    for a in range(2, 10 + 1):
        result = discrete_log(a, n, b)
        if result == 0:
            unexist += 1
            continue
        
        a_sum += result
    
    sum_result += a_sum
end = perf_counter()

print(f"Nilai logaritma diskrit = {sum_result}")
print(f"Di antara nilai logaritma diskrit yang dihitung, terdapat {unexist} di antaranya yang nilainya tidak eksis.")
print(f"\nWaktu eksekusi: {end - start: .4f}s")


Nilai logaritma diskrit = 8101199
Di antara nilai logaritma diskrit yang dihitung, terdapat 32486 di antaranya yang nilainya tidak eksis.

Waktu eksekusi:  5377.3644s


*Documentation:*

Nilai logaritma diskrit = 8101199
Di antara nilai logaritma diskrit yang dihitung, terdapat 32486 di antaranya yang nilainya tidak eksis.

Waktu eksekusi:  5377.3644s

### $GF(2^8)$

#### Polinomial $\times$ Monomial

In [1]:
def GF_multiply_by_x(P: str) -> str:
    if P[0] == "0":
        return P[1:] + "0"
    else:
        P = P[1:] + "0"
        constant = "00011011"
        
        result = ""
        for i in range(len(P)):
            result += str((int(P[i]) + int(constant[i])) % 2)

        return result


print(GF_multiply_by_x("01010011") == "10100110")
print(GF_multiply_by_x("11010011") == "10111101")


True
True


#### Polinomial $\times$ Monomial $x^k$

In [2]:
def GF_multiply_by_x_k(P: str, k: int) -> str:
    new = P
    for i in range(k):
        new = GF_multiply_by_x(new)
        
    return new


# GF_multiply_by_x_k("01010011", 3) == "10101110"
GF_multiply_by_x_k("01010011", 1) == "10100110"


True

#### Polinomial $\times$ Polinomial


In [3]:
def GF_multiply(P: str, Q: str) -> str:
    l = []
    l.append("0"*8) if Q[-1] == "0" else l.append(P)
    
    for i in range(len(P)-2, -1, -1):
        if Q[i] != "0":
            l.append(GF_multiply_by_x_k(P, abs(i-7)))
        else:
            l.append("0"*8)

    result = ""
    for j in range(len(l)):
        dumb = 0
        for k in range(8):
            dumb += int(l[k][j])
            dumb = dumb % 2
        result += str(dumb)

    return result

    
GF_multiply("01010011", "00001011") == "01011011"


True

#### QUIZ

In [7]:
# (0F)^200
nol_f_200 = "00001111"
for i in range(200 - 1):
    nol_f_200 = GF_multiply(nol_f_200, "00001111")
# print(f"(0F)^200: {nol_f_200}")


# (BE)^100
be_100 = "10111110"
for i in range(100 - 1):
    be_100 = GF_multiply(be_100, "10111110")
# print(f"(BE)^100: {be_100}")


# (0F)^200 * (BE)^100
ans_ = GF_multiply(nol_f_200, be_100)
print(f"(0F)^200 * (BE)^100: {ans_}")


# (DA)^50
da_50 = "11011010"
for i in range(50 - 1):
    da_50 = GF_multiply(da_50, "11011010")
# print(f"(DA)^50: {da_50}")


# (0F)^200 * (DA)^50
ans__ = GF_multiply(nol_f_200, da_50)
print(f"(0F)^200 * (BE)^50: {ans__}")


(0F)^200 * (BE)^100: 01100001
(0F)^200 * (BE)^50: 10111101


01100001 + 10111101 = 11011100 = "DC"